### ISSUE TITLE
Implicit ratings

### ISSUE THREAD

**Issue by micheledemeo** [massquantity/LibRecommender#16]
Hi all,
I have a a couple of questions about the possibility of using some algorithms (available on LibRecommender) for implicit ratings.

For example let's talk about using SVD++ with task="ranking" on sample_movielens_rating.dat:
1. Can someone confirm that the assumption here is that the missing entries, that is the not rated items, are unknown? 
1. It seems to me that the Total Loss is minimized on the whole rating matrix, where not-rated-items (for every user i) are filled with 0. Is it correct?

Thank you for building up this library :)

**Comment by massquantity** [comment]
Yes, your understanding is basically correct. If you use task="ranking", the library assumes all positive samples are labeled 1, and negative(not rated) samples are labeled as 0. There are generally two scenarios:

1.. If you have a data that only contains positive samples, such as ratings in  `sample_movielens_rating.dat` , you need to sample negative samples from the whole item pool first. Because it's not feasible to use all not-rated-items to train the model. 

```python
train_data, data_info = DatasetPure.build_trainset(...)
test_data = DatasetPure.build_testset(...)
train_data.build_negative_samples(data_info, num_neg=1)  # change the value of num_neg if you want to sample more than one negative sample per positive sample
test_data.build_negative_samples(data_info, num_neg=1)
```
2.. If you have a data that contains both negative and positive samples and the labels are not 1 and 0, you should manually change the labels of all positive samples to 1 and labels of negative samples to 0. But in this case there is no need to do negative sampling.

By the way, I think the main cause of confusion is the lack of documentation, which I always wanted to but didn't have the time to do...

**Comment by micheledemeo** [comment]
Ok thank you for the clear explanation 👍, indeed my intention was to argue that not rated items are generally different from negative rating. If you never saw a movie, it doesn't mean you don't like it.
It's also true that reasonably robust solutions can be obtained by treating the missing entries as 0 in the modeling process. This is especially the case when the prior probability of an entry to be 0 is very large: for instance, I would typically never buy the vast majority of items on Amazon.

On the other hand, it's possible to leverage the Latent Matrix Factorization with missing ratings, where the Loss function is actually minimized on rated items only. However, I'm not saying you did something wrong :) , I was asking a confirmation on your implementation of SDV, where you're using the approach of filling missing ratings with 0.

Thank you for sharing you work

**Comment by massquantity** [comment]
Yeah I agree with you, it is generally not appropriate to assume all not rated items are negative ones.
By saying "Loss function is actually minimized on rated items only", did you mean the method described in the paper [Collaborative Filtering for Implicit Feedback Datasets](http://yifanhu.net/PUB/cf.pdf) ？  This is all I can think of..

**Comment by micheledemeo** [comment]
Yes this is what I meant:
![image](https://user-images.githubusercontent.com/7612796/96692569-7a19f900-1386-11eb-87ec-54146162c2f1.png)
